pmlc/dialect/abi/ir/ops.td

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 Intel Corporation
+
+#ifndef __PML_ABI_OPS__
+#define __PML_ABI_OPS__
+
+include "mlir/Dialect/StandardOps/IR/StandardOpsBase.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+
+include "pmlc/util/enums.td"
+
+def ABI_Dialect : Dialect {
+  let name = "abi";
+  let cppNamespace = "pmlc::dialect::abi";
+}
+
+class ABI_Op<string mnemonic, list<OpTrait> traits = []> :
+    Op<ABI_Dialect, mnemonic, traits>;
+
+def ABI_LoopOp : ABI_Op<"loop", [
+  IsolatedFromAbove, HasParent<"ModuleOp">, RecursiveSideEffects]> {
+  let summary = "Repeat runs of a network";
+  let description = [{
+    The "abi.loop" operation represents repeated runs of a program,
+    taking new inputs and generating new outputs each time.
+    This allows loop-invariant code motion to move setup and teardown
+    operations outside of the loop, so that their cost is ammortized over
+    many runs of the program.
+  }];
+  let regions = (region AnyRegion:$initRegion, AnyRegion:$bodyRegion, AnyRegion:$finiRegion);
```

Comment:
Hmm, why multiple regions? Seems like we want one region and have the code be hoisted to either pre- or post- this op.

Reply:
Several reasons:

1. This approach makes the abi-to-functions pass simpler, since there's a direct mapping from regions to functions, the region arguments correspond to the function arguments, &c -- no need to go sniffing around to find the loop.

1. This approach is also automatically correct-by-construction: an abi.loop is an isolated-from-above child of a ModuleOp; there's no correct way to give it any other parent, to have multiple abi.loop operations within a single function, &c.

1. This approach makes the semantics of what's going on more clear: we're recasting a function into a separate initialization, execution, and finalization section, at the top level, and then transforming it from there.  The alternative, having a single function wrapping a loop, still "looks like" a single function that you call and get some result from; it's not obvious that additional lowering is needed in order to match the ABI expected by JitExecutable.

Reply:
Wait, so this is intended to replace a `func` op?

Reply:
Yup.  The conversion takes the body of a `func`, puts it into an `abi.loop`'s body region, sets up arguments, and calls it a day.

I'm thinking the next PR in this series will be that conversion pass, with a lit test to show what's going on -- think I should combine it with the current PR?

Reply:
Yeah, I also didn't catch that, although it's all in the PR (HasParent<"ModuleOp"> for example).  Question: Does this design actually work with LICM?  I realize there is some concerns on 'destructor' motion, so we might need our own transforms anyway, but I'd like LICM to work if possible.

Reply:
I agree that we should be able to use the standard LICM pass. If we need to handle the split of alloc/dealloc, I think that's a separate thing that we can address in some other pass.

Reply:
I'm open to the idea of making abi.loop be the primary entry point and having it live directly in the module, but I do worry a bit about losing all the functionality that exists on FuncOp.

Reply:
Agreed, how to we represent attributes on arguments? This is something FuncOp supports. Is there a FunctionLike interface or trait that we can use to acquire the features we want from standard FuncOp?

Reply:
It does not work with the standard LICM pass as-is -- I was saving adding LoopLikeInterface for a subsequent PR (although I do have it implemented.

Also, as @flaub notes, LICM doesn't handle alloc/dealloc; it also doesn't handle anything with memory side-effects.  I wound up adding my own pass, along with an interface to add support for "move operations to epilog".

TL;DR: Later.

Reply:
What's the use case for attributes on arguments at this stage of the lowering?  (NB This is pretty close to the point where we're going to be converting to `llvm.func` and generating code.)

Reply:
... also, note that `FunctionLike` implies a single region, a set of arguments and a set of results, &c; we might want to add argument attribute support, but this op isn't `FunctionLike`.

Reply:
Ok on further thought here is my feelings on LICM: I think we could make most (all?) things work using LICM via cleverness (i.e. making allocs no-side-effect and adding deallocs to fini only after we are done, after all, we already begin with allocs only and add deallocs later).  However, this is sometimes awkward since really we want 'idempotent' ops to be moved to do-once, not just 'no-side-effect' exactly, and things like 'alloc' actually *are* marked as side-effect, so even if we wanted to use LICM, we would need a special version of alloc.  Moreover, LICM is < 130 lines of code (including a lot of comments and boilerplate), so it's not like partially duplicating bits of it is a major engineering cost.

Conclusion, reuse is probably not worth it here.  At that point, the issue of 1 region vs 3 regions becomes less relevant from that perspective, and there is definitely something nice about having abi.loop be top level in the module.

Reply:
FWIW: LICM works on arbitrary LoopLike ops, and uses the op for manipulations, so the multiple regions and top-level aspect aren't incompatibilities at all; I used LICM early on, and only implemented a separate hoisting pass to handle ops with side-effects and where I needed move-to-epilog semantics.

But I still (clearly :-) like having abi.loop as a top-level op corresponding to "How we're going to run this network".